strawberryfields/program.py

```diff
@@ -0,0 +1,897 @@
+    several :class:`Program` instances.
+
+    Args:
+        op (~strawberryfields.ops.Operation): quantum operation to apply
+        reg (Sequence[RegRef]): Subsystems to which the operation is applied.
+            Note that the order matters here.
+    """
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, op, reg, decomp=False):
+        # accept a single RegRef in addition to a Sequence
+        if not isinstance(reg, Sequence):
+            reg = [reg]
+
+        #: Operation: quantum operation to apply
+        self.op = op
+        #: Sequence[RegRef]: subsystems to which the operation is applied
+        self.reg = reg
+        #: bool: is this Command a part of a decomposition?
+        self.decomp = decomp
+
+    def __str__(self):
+        """Print the command using Blackbird syntax."""
+        temp = str(self.op)
+        if self.op.ns == 0:
+            # op takes no subsystems as parameters, do not print anything more
+            return temp
+        return '{} | ({})'.format(temp, ", ".join([str(rr) for rr in self.reg]))
+
+    def get_dependencies(self):
+        """Subsystems the command depends on.
+
+        Combination of ``self.reg`` and ``self.op.extra_deps``.
+
+        .. note:: ``extra_deps`` are used to ensure that the measurement
+            happens before the result is used, but this is a bit too strict:
+            two gates depending on the same measurement result but otherwise
+            acting on different subsystems should commute.
+
+        Returns:
+            set[RegRef]: set of subsystems the command depends on
+        """
+        deps = self.op.extra_deps | set(self.reg)
+        return deps
+
+
+class RegRef:
+    """Quantum register reference.
+
+    The objects of this class refer to a specific subsystem (mode) of
+    a quantum register.
+
+    Within the scope of each :class:`Program` instance, only one RegRef instance
+    should exist per subsystem. :class:`Program` keeps the authoritative mapping
+    of subsystem indices to RegRef instances.
+    Subsystem measurement results are stored in the "official" RegRef object.
+    If other RegRefs objects referring to the same subsystem exist, they will
+    not be updated. Once a RegRef is assigned a subsystem index it will never
+    change, not even if the subsystem is deleted.
+
+    The RegRefs are constructed in :meth:`Program._add_subsystems`.
+
+    Args:
+        ind (int): index of the register subsystem referred to
+    """
+    # pylint: disable=too-few-public-methods
+
+    def __init__(self, ind):
+        self.ind = ind   #: int: subsystem index
+        self.val = None  #: Real: measured eigenvalue. None if the subsystem has not been measured yet
+        self.active = True  #: bool: True at construction, False after the subsystem is deleted
+
+    def __str__(self):
+        return 'q[{}]'.format(self.ind)
+
+    def __hash__(self):
+        return hash((self.ind, self.active))
```

Comment:
haha the temptation to use `__eq__` was too much :laughing:

Reply:
Yeah, this way I can compare two dictionaries with ==. I'll add the missing docstring.